File: LeetCode/Medium/0208-implement-trie-prefix-tree/0208-implement-trie-prefix-tree.py
class Trie:
    class TrieNode:
        def __init__(self):
            self.children = {}
            self.is_end_of_the_word = False

    def __init__(self):
        self.root = Trie.TrieNode()

    def insert(self, word: str) -> None:
        current = self.root
        for char in word:
            if char not in current.children:
                current.children[char] = Trie.TrieNode()
            current = current.children[char]
        current.is_end_of_the_word = True
        
    def _search(self, string: str, finisher: Callable[[TrieNode], bool]):
        current = self.root
        for char in string:
            if char not in current.children:
                return False
            current = current.children[char]
        return finisher(current)

    def search(self, word: str) -> bool:
        return self._search(word, lambda node: node.is_end_of_the_word)

    def startsWith(self, prefix: str) -> bool:
        return self._search(prefix, lambda _: True)
        


# Your Trie object will be instantiated and called as such:
# obj = Trie()
# obj.insert(word)
# param_2 = obj.search(word)
# param_3 = obj.startsWith(prefix)

File: LeetCode/Medium/0208-implement-trie-prefix-tree/0208-implement-trie-prefix-tree.cs
public class Trie 
{
    private static int AlphabetLettersCount = 26;
    private TrieNode root = new TrieNode();
    
    public void Insert(string word) {
        var current = root;
        foreach (var c in word)
        {
            if (current.Children == null || current.Children[c - 'a'] == null)
            {
                if (current.Children == null) current.Children = new TrieNode[AlphabetLettersCount]; 
                current.Children[c - 'a'] = new TrieNode();
            }
            current = current.Children[c - 'a'];
        }

        current.IsEndOfTheWord = true;
    }
    
    public bool Search(string word) {
        var node = GetEndNodeOf(word);
        return node != null && node.IsEndOfTheWord;
    }
    
    public bool StartsWith(string prefix) {
        return GetEndNodeOf(prefix) != null;
    }

    private TrieNode GetEndNodeOf(string str)
    {
        var current = root;
        foreach (var c in str)
        {
            if (current.Children == null || current.Children[c - 'a'] == null) return null;
            current = current.Children[c - 'a'];
        }

        return current;
    }

    class TrieNode
    {
        public bool IsEndOfTheWord = false;
        public TrieNode[] Children;
    }
}


/**
 * Your Trie object will be instantiated and called as such:
 * Trie obj = new Trie();
 * obj.Insert(word);
 * bool param_2 = obj.Search(word);
 * bool param_3 = obj.StartsWith(prefix);
 */